<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Costa, W. G." />

<meta name="date" content="2022-10-27" />

<title>Genomic selection in a single environment</title>

<script src="site_libs/header-attrs-2.17/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/flatly.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; } /* Alert */
code span.an { color: #008000; } /* Annotation */
code span.at { } /* Attribute */
code span.bu { } /* BuiltIn */
code span.cf { color: #0000ff; } /* ControlFlow */
code span.ch { color: #008080; } /* Char */
code span.cn { } /* Constant */
code span.co { color: #008000; } /* Comment */
code span.cv { color: #008000; } /* CommentVar */
code span.do { color: #008000; } /* Documentation */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.im { } /* Import */
code span.in { color: #008000; } /* Information */
code span.kw { color: #0000ff; } /* Keyword */
code span.op { } /* Operator */
code span.ot { color: #ff4000; } /* Other */
code span.pp { color: #ff4000; } /* Preprocessor */
code span.sc { color: #008080; } /* SpecialChar */
code span.ss { color: #008080; } /* SpecialString */
code span.st { color: #008080; } /* String */
code span.va { } /* Variable */
code span.vs { color: #008080; } /* VerbatimString */
code span.wa { color: #008000; font-weight: bold; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Genomic Selection Course</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="genomicselection.html">Genomic Selection Single Environment</a>
</li>
<li>
  <a href="genomicselection2.html">Genomic Selection Multi Environment</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Genomic selection in a single
environment</h1>
<h4 class="author">Costa, W. G.<a href="#fn1" class="footnote-ref"
id="fnref1"><sup>1</sup></a></h4>
<h4 class="date">2022-10-27</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2022-10-27
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>Genomic-Selection_Course/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has staged changes. To know which version of the
R Markdown file created these results, you’ll want to first commit it to
the Git repo. If you’re still working on the analysis, you can ignore
this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20221026code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20221026)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20221026code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20221026)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomWevertonGomesCostaGenomicSelectionCoursetree4f64fdfa265df3c1548088520eff8dce0bffc461targetblank4f64fdfa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/WevertonGomesCosta/Genomic-Selection_Course/tree/4f64fdfa265df3c1548088520eff8dce0bffc461" target="_blank">4f64fdf</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomWevertonGomesCostaGenomicSelectionCoursetree4f64fdfa265df3c1548088520eff8dce0bffc461targetblank4f64fdfa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/WevertonGomesCosta/Genomic-Selection_Course/tree/4f64fdfa265df3c1548088520eff8dce0bffc461" target="_blank">4f64fdf</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/

Untracked files:
    Untracked:  ETA_1_b.dat
    Untracked:  ETA_2_varB.dat
    Untracked:  ETA_2_varU.dat
    Untracked:  ETA_3_varB.dat
    Untracked:  ETA_3_varU.dat
    Untracked:  analysis/genomicselection2.Rmd

Unstaged changes:
    Modified:   analysis/_site.yml
    Modified:   analysis/genomicselection.Rmd
    Modified:   analysis/license.Rmd
    Modified:   mu.dat
    Modified:   varE.dat

Staged changes:
    New:        ETA_1_parBayesB.dat
    New:        ETA_1_varB.dat
    New:        ETA_1_varU.dat
    Modified:   analysis/about.Rmd
    New:        analysis/genomicselection.Rmd
    Modified:   analysis/index.Rmd
    New:        mu.dat
    New:        varE.dat

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
There are no past versions. Publish this analysis with
<code>wflow_publish()</code> to start tracking its development.
</p>
<hr>
</div>
</div>
</div>
<p>In this document you will find a step-by-step of how to perform the
genomic selection in a single environment.</p>
<p>To perform the analyses, we will need the following packages:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(kableExtra)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(RCurl)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(data.table)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(AGHmatrix)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(rrBLUP)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(BGLR)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(tidyverse)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(ComplexHeatmap)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(cvTools)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">require</span>(patchwork)</span></code></pre></div>
<p><br></p>
<div id="data" class="section level2">
<h2>Data</h2>
<p>Phenotypic adjusted means and genomic data are available online at
this <a
href="https://github.com/samuelbfernandes/Trait-assisted-GS">GitHub
page</a>, which refers to the article of <span
class="citation">@fernandes_efficiency_2018</span>.</p>
<p>We can load the adj. means data directly from the Github using the
following command:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>adjmeans <span class="ot">=</span> <span class="fu">read.csv</span>(<span class="st">&quot;https://raw.githubusercontent.com/samuelbfernandes/Trait-assisted-GS/master/means.csv&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>adjmeans<span class="sc">$</span>GENO <span class="ot">=</span> <span class="fu">as.factor</span>(adjmeans<span class="sc">$</span>GENO)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>adjmeans<span class="sc">$</span>LOC <span class="ot">=</span> <span class="fu">as.factor</span>(adjmeans<span class="sc">$</span>LOC)</span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(adjmeans,<span class="dv">10</span>) <span class="sc">%&gt;%</span> <span class="fu">kbl</span>(<span class="at">escape =</span> F, <span class="at">align =</span> <span class="st">&#39;c&#39;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_classic</span>(<span class="st">&quot;hover&quot;</span>,<span class="at">full_width =</span> T, <span class="at">position=</span><span class="st">&quot;center&quot;</span>, <span class="at">fixed_thead =</span> T)</span></code></pre></div>
<table class=" lightable-classic lightable-hover" style="font-family: &quot;Arial Narrow&quot;, &quot;Source Sans Pro&quot;, sans-serif; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
X
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
LOC
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
GENO
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
h1
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
h2
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
h3
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
h4
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
M
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
Y
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
Belko2
</td>
<td style="text-align:center;">
0.1710
</td>
<td style="text-align:center;">
1.050
</td>
<td style="text-align:center;">
1.975
</td>
<td style="text-align:center;">
2.875
</td>
<td style="text-align:center;">
0.7235394
</td>
<td style="text-align:center;">
9.120230
</td>
</tr>
<tr>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
CHRFS9
</td>
<td style="text-align:center;">
0.2790
</td>
<td style="text-align:center;">
1.700
</td>
<td style="text-align:center;">
2.775
</td>
<td style="text-align:center;">
3.550
</td>
<td style="text-align:center;">
0.6080388
</td>
<td style="text-align:center;">
11.384432
</td>
</tr>
<tr>
<td style="text-align:center;">
3
</td>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
CHRSS2
</td>
<td style="text-align:center;">
0.2900
</td>
<td style="text-align:center;">
1.825
</td>
<td style="text-align:center;">
2.600
</td>
<td style="text-align:center;">
3.850
</td>
<td style="text-align:center;">
0.5748191
</td>
<td style="text-align:center;">
15.087747
</td>
</tr>
<tr>
<td style="text-align:center;">
4
</td>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
E105
</td>
<td style="text-align:center;">
0.2520
</td>
<td style="text-align:center;">
1.275
</td>
<td style="text-align:center;">
2.375
</td>
<td style="text-align:center;">
3.450
</td>
<td style="text-align:center;">
0.6954704
</td>
<td style="text-align:center;">
10.373466
</td>
</tr>
<tr>
<td style="text-align:center;">
5
</td>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
Epo
</td>
<td style="text-align:center;">
0.2610
</td>
<td style="text-align:center;">
1.200
</td>
<td style="text-align:center;">
2.325
</td>
<td style="text-align:center;">
3.325
</td>
<td style="text-align:center;">
0.7139562
</td>
<td style="text-align:center;">
6.726541
</td>
</tr>
<tr>
<td style="text-align:center;">
6
</td>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
ES5200
</td>
<td style="text-align:center;">
0.3560
</td>
<td style="text-align:center;">
1.800
</td>
<td style="text-align:center;">
2.600
</td>
<td style="text-align:center;">
3.725
</td>
<td style="text-align:center;">
0.6593198
</td>
<td style="text-align:center;">
12.999241
</td>
</tr>
<tr>
<td style="text-align:center;">
7
</td>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
Grassl
</td>
<td style="text-align:center;">
0.2465
</td>
<td style="text-align:center;">
1.550
</td>
<td style="text-align:center;">
2.525
</td>
<td style="text-align:center;">
3.500
</td>
<td style="text-align:center;">
0.6991563
</td>
<td style="text-align:center;">
11.971261
</td>
</tr>
<tr>
<td style="text-align:center;">
8
</td>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
GS109
</td>
<td style="text-align:center;">
0.2280
</td>
<td style="text-align:center;">
1.275
</td>
<td style="text-align:center;">
2.175
</td>
<td style="text-align:center;">
2.900
</td>
<td style="text-align:center;">
0.6784228
</td>
<td style="text-align:center;">
8.832171
</td>
</tr>
<tr>
<td style="text-align:center;">
9
</td>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
Kalaicou
</td>
<td style="text-align:center;">
0.2325
</td>
<td style="text-align:center;">
1.100
</td>
<td style="text-align:center;">
2.200
</td>
<td style="text-align:center;">
3.025
</td>
<td style="text-align:center;">
0.7265696
</td>
<td style="text-align:center;">
7.975786
</td>
</tr>
<tr>
<td style="text-align:center;">
10
</td>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
MA38
</td>
<td style="text-align:center;">
0.2970
</td>
<td style="text-align:center;">
1.375
</td>
<td style="text-align:center;">
2.400
</td>
<td style="text-align:center;">
3.050
</td>
<td style="text-align:center;">
0.6855849
</td>
<td style="text-align:center;">
11.769430
</td>
</tr>
</tbody>
</table>
<p><br></p>
<p>The genomic data can also be loaded directly:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>SNPs <span class="ot">=</span> <span class="fu">fread</span>(<span class="st">&quot;https://raw.githubusercontent.com/samuelbfernandes/Trait-assisted-GS/master/snps.csv&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>names_lines <span class="ot">=</span> SNPs[,<span class="dv">1</span>]</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>SNPs <span class="ot">=</span> SNPs[,<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>SNPs <span class="ot">=</span> <span class="fu">as.matrix</span>(SNPs[<span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(SNPs)[<span class="dv">1</span>], <span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(SNPs)[<span class="dv">2</span>]])</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(SNPs) <span class="ot">=</span> names_lines<span class="sc">$</span>V1</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(SNPs)</span></code></pre></div>
<pre><code>[1]   453 58960</code></pre>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>SNPs[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>         X1_23664 X1_26742 X1_33900 X1_62283 X1_62548
E105           -1       -1        1       -1        1
Epo            -1       -1       -1       -1        0
Grassl         -1        1       -1       -1        1
GS109           1       -1       -1        1        1
Kalaicou        1       -1       -1        1        1</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>SNPs[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>] <span class="sc">%&gt;%</span> <span class="fu">kbl</span>(<span class="at">escape =</span> F, <span class="at">align =</span> <span class="st">&#39;c&#39;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_classic</span>(<span class="st">&quot;hover&quot;</span>,<span class="at">full_width =</span> T, <span class="at">position=</span><span class="st">&quot;center&quot;</span>, <span class="at">fixed_thead =</span> T) <span class="sc">%&gt;%</span> <span class="fu">footnote</span>(<span class="st">&quot;Dimension: 453 $</span><span class="sc">\\</span><span class="st">times$ 58960&quot;</span>,<span class="at">general_title =</span> <span class="st">&quot;&quot;</span>)</span></code></pre></div>
<table class=" lightable-classic lightable-hover" style="font-family: &quot;Arial Narrow&quot;, &quot;Source Sans Pro&quot;, sans-serif; margin-left: auto; margin-right: auto;border-bottom: 0;">
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
X1_23664
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
X1_26742
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
X1_33900
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
X1_62283
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
X1_62548
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
E105
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
Epo
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
Grassl
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
GS109
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
Kalaicou
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
-1
</td>
<td style="text-align:center;">
1
</td>
<td style="text-align:center;">
1
</td>
</tr>
</tbody>
<tfoot>
<tr>
<td style="padding: 0; " colspan="100%">
<sup></sup> Dimension: 453 <span class="math inline">\(\times\)</span>
58960
</td>
</tr>
</tfoot>
</table>
<p><br></p>
<p><br></p>
</div>
<div id="subsetting" class="section level2">
<h2>Subsetting</h2>
<p>Note that we have three levels in the “LOC” factor:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>adjmeans <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(LOC) <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(<span class="fu">length</span>(GENO)) <span class="sc">%&gt;%</span> <span class="fu">kbl</span>(<span class="at">escape =</span> F, <span class="at">align =</span> <span class="st">&#39;c&#39;</span>, <span class="at">col.names =</span> <span class="fu">c</span>(<span class="st">&quot;Environments&quot;</span>,<span class="st">&quot;Number of genotypes&quot;</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_classic</span>(<span class="st">&quot;hover&quot;</span>,<span class="at">full_width =</span> F, <span class="at">position=</span><span class="st">&quot;center&quot;</span>, <span class="at">fixed_thead =</span> T)</span></code></pre></div>
<table class=" lightable-classic lightable-hover" style="font-family: &quot;Arial Narrow&quot;, &quot;Source Sans Pro&quot;, sans-serif; width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
Environments
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
Number of genotypes
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:center;">
12
</td>
<td style="text-align:center;">
192
</td>
</tr>
<tr>
<td style="text-align:center;">
13
</td>
<td style="text-align:center;">
438
</td>
</tr>
<tr>
<td style="text-align:center;">
14
</td>
<td style="text-align:center;">
431
</td>
</tr>
</tbody>
</table>
<p>Therefore, we need to choose only one to perform the analysis. We
will choose the LOC “13”:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>adjmeans_L <span class="ot">=</span> <span class="fu">droplevels</span>(<span class="fu">subset</span>(adjmeans, LOC <span class="sc">==</span> <span class="dv">13</span>))</span></code></pre></div>
<p>Below, we can see the distribution of frequencies regarding the grain
yield trait in LOC “13”. Note that the similarity to the gaussian
distribution:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(adjmeans_L, <span class="fu">aes</span>(<span class="at">x =</span> Y)) <span class="sc">+</span>  </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">y =</span> ..density..), <span class="at">bins =</span> <span class="dv">30</span>, </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">colour =</span> <span class="st">&quot;black&quot;</span>, <span class="at">fill =</span> <span class="st">&quot;steelblue&quot;</span>) <span class="sc">+</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">alpha =</span> .<span class="dv">7</span>, <span class="at">size =</span> <span class="fl">1.5</span>, <span class="at">colour =</span> <span class="st">&quot;tomato&quot;</span>) <span class="sc">+</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="cn">NULL</span>, <span class="at">title =</span> <span class="st">&quot;Grain Yield&quot;</span>,<span class="at">y=</span><span class="st">&quot;Density&quot;</span>)</span></code></pre></div>
<p><img src="figure/genomicselection.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="building-the-g-matrix" class="section level2">
<h2>Building the G matrix</h2>
<p>Note that the SNP matrix is coded using 1 and -1 for homozygotes and
0 for heterozygotes. We may change this codification to a dosage
solution. In other words, we can represent the dosage of a reference
allele (A, for example) to code the loci. If the locus has AA, then we
will attribute the number 2. If it is a heterozygote, i.e. have only one
A, it will be 1. Otherwise, the locus will be represented by 0:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>SNPs <span class="ot">=</span> SNPs <span class="sc">+</span> <span class="dv">1</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>SNPs[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>         X1_23664 X1_26742 X1_33900 X1_62283 X1_62548
E105            0        0        2        0        2
Epo             0        0        0        0        1
Grassl          0        2        0        0        2
GS109           2        0        0        2        2
Kalaicou        2        0        0        2        2</code></pre>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>SNPs[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>] <span class="sc">%&gt;%</span> <span class="fu">kbl</span>(<span class="at">escape =</span> F, <span class="at">align =</span> <span class="st">&#39;c&#39;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kable_paper</span>(<span class="st">&quot;hover&quot;</span>,<span class="at">full_width =</span> T, <span class="at">position=</span><span class="st">&quot;center&quot;</span>, <span class="at">fixed_thead =</span> T) </span></code></pre></div>
<table class=" lightable-paper lightable-hover" style="font-family: &quot;Arial Narrow&quot;, arial, helvetica, sans-serif; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
X1_23664
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
X1_26742
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
X1_33900
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
X1_62283
</th>
<th style="text-align:center;position: sticky; top:0; background-color: #FFFFFF;">
X1_62548
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
E105
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Epo
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
1
</td>
</tr>
<tr>
<td style="text-align:left;">
Grassl
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
GS109
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Kalaicou
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
0
</td>
<td style="text-align:center;">
2
</td>
<td style="text-align:center;">
2
</td>
</tr>
</tbody>
</table>
<p>To simplify the analysis, we will subset the SNP matrix and use only
5000 SNPs:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>SNPs <span class="ot">=</span> SNPs[,<span class="dv">1</span><span class="sc">:</span><span class="dv">5000</span>]</span></code></pre></div>
<p>We will use the <a
href="https://cran.r-project.org/web/packages/AGHmatrix/vignettes/Tutorial_AGHmatrix.html">AGHmatrix</a>
package <span class="citation">[@amadeu_aghmatrix_2016]</span> to build
the G matrix:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>G_matrix <span class="ot">=</span> <span class="fu">Gmatrix</span>(SNPs, <span class="at">method =</span> <span class="st">&quot;VanRaden&quot;</span>, <span class="at">ploidy =</span> <span class="dv">2</span>, <span class="at">missingValue =</span> <span class="cn">NA</span>)</span></code></pre></div>
<pre><code>Initial data: 
    Number of Individuals: 453 
    Number of Markers: 5000 

Missing data check: 
    Total SNPs: 5000 
     0 SNPs dropped due to missing data threshold of 1 
    Total of: 5000  SNPs 
MAF check: 
    No SNPs with MAF below 0 
Monomorphic check: 
    No monomorphic SNPs 
Summary check: 
    Initial:  5000 SNPs 
    Final:  5000  SNPs ( 0  SNPs removed) 
 
Completed! Time = 4.88  seconds </code></pre>
<p>Now we have a square genomic relationship matrix whose number of rows
and columns is equal to the number of individuals.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(G_matrix)</span></code></pre></div>
<pre><code>[1] 453 453</code></pre>
<p>We can represent this matrix using a heatmap:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">Heatmap</span>(G_matrix, <span class="at">show_row_names =</span> F, <span class="at">show_column_names =</span> F,</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">heatmap_legend_param =</span> <span class="fu">list</span>(<span class="at">title =</span> <span class="st">&quot;Res&quot;</span>))</span></code></pre></div>
<p><img src="figure/genomicselection.Rmd/unnamed-chunk-16-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>“Res” in the heatmap legend title is for “Resemblance”.</p>
<p><br></p>
</div>
<div id="rrblup" class="section level2">
<h2>RRBLUP</h2>
<p>The Ridge Regression BLUP, or RRBLUP, will predict the marker effect.
In the RRBLUP, we will use the matrix of markers directly. For this
purpose, we will use only individuals with adj. means and SNPs
available.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Selecting the adjmeanstypes with SNP information</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">levels</span>(adjmeans_L<span class="sc">$</span>GENO))</span></code></pre></div>
<pre><code>[1] 438</code></pre>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>adjmeans_L <span class="ot">=</span> <span class="fu">droplevels</span>(adjmeans_L[adjmeans_L<span class="sc">$</span>GENO <span class="sc">%in%</span> <span class="fu">rownames</span>(SNPs), ])</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">levels</span>(adjmeans_L<span class="sc">$</span>GENO))</span></code></pre></div>
<pre><code>[1] 415</code></pre>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filtering the SNPs matrix: only genotyped and adjmeanstyped</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>index <span class="ot">=</span> <span class="fu">rownames</span>(SNPs) <span class="sc">%in%</span> adjmeans_L<span class="sc">$</span>GENO</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>SNPs_L <span class="ot">=</span> SNPs[index,]</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(SNPs_L)</span></code></pre></div>
<pre><code>[1]  415 5000</code></pre>
<p>Note that 38 genotypes were dropped from the SNPs matrix and the data
containing the means. Those did not have both means and genomic
information.</p>
<p>Now that we filtered our data, we can employ the RRBLUP. For this
purpose, we will use the <a
href="https://cran.r-project.org/web/packages/rrBLUP/rrBLUP.pdf">rrBLUP</a>
package <span class="citation">[@endelman_2011]</span>. In the code
below, <em>y</em> is for the vector with the means, <em>Z</em> is where
we will insert the SNPs matrix, <em>K</em> is for a covariance matrix
for the random effects, which will be and identity matrix by default;
and <em>X</em> is a design matrix for the fixed effects, which will be a
vector of ones (<strong>1</strong>) by default. Note that we are
returning to the “1, 0, -1” codification in the SNPs matrix. This is a
requirement of the <em>rrBlUP</em> package.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>RRBLUP <span class="ot">=</span> <span class="fu">mixed.solve</span>(<span class="at">y =</span> adjmeans_L<span class="sc">$</span>Y, <span class="at">Z =</span> SNPs_L<span class="dv">-1</span>, <span class="at">K =</span> <span class="cn">NULL</span>, <span class="at">X =</span> <span class="cn">NULL</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">as.data.frame</span>(RRBLUP<span class="sc">$</span>u) <span class="sc">%&gt;%</span> </span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">&quot;SNP&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">BLUP =</span> <span class="st">&#39;RRBLUP$u&#39;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> SNP, <span class="at">y =</span> BLUP))<span class="sc">+</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="figure/genomicselection.Rmd/unnamed-chunk-18-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>The scatter plot above represents the additive genetic value of each
marker. Once we have acquired these values, we may calculate the Genomic
Estimated Breeding Values (GEBV) of the genotypes. These are the product
of the SNPs matrix with the vector of the markers’ genetic values:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>GEBV <span class="ot">=</span> SNPs_L <span class="sc">%*%</span> RRBLUP<span class="sc">$</span>u</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="fu">as.data.frame</span>(GEBV) <span class="sc">%&gt;%</span> </span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">&quot;Genotype&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">GEBV =</span> V1) <span class="sc">%&gt;%</span> </span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> Genotype, <span class="at">y =</span> GEBV))<span class="sc">+</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="figure/genomicselection.Rmd/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Then, we can estimate the correlation between the GEBV and the
adjusted means.</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>acc_RRBLUP <span class="ot">=</span> <span class="fu">cor</span>(adjmeans_L<span class="sc">$</span>Y, GEBV, <span class="at">use =</span> <span class="st">&quot;complete.obs&quot;</span>);acc_RRBLUP</span></code></pre></div>
<pre><code>          [,1]
[1,] 0.2896504</code></pre>
<p><br></p>
</div>
<div id="gblup" class="section level2">
<h2>GBLUP</h2>
<p>In the GBLUP, we will use the G matrix instead of the SNPs matrix.
Thus, we will obtain the GEBV directly. Note that we will need to build
the G matrix again, since some genotypes were dropped after our
filtering. The rrBLUP package has a function called “A.mat” that build
the Additive Genomic Matrix from a SNP matrix with “-1,0,1”
codification:</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>GBLUP <span class="ot">=</span> <span class="fu">mixed.solve</span>(adjmeans_L<span class="sc">$</span>Y, <span class="at">K =</span> <span class="fu">A.mat</span>(SNPs_L<span class="dv">-1</span>))</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">as.data.frame</span>(GBLUP<span class="sc">$</span>u) <span class="sc">%&gt;%</span> </span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">&quot;Genotype&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">BLUP =</span> <span class="st">&#39;GBLUP$u&#39;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> Genotype, <span class="at">y =</span> BLUP))<span class="sc">+</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>())</span></code></pre></div>
<p><img src="figure/genomicselection.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>As in the RRBLUP, we may calculate the accuracy of prediction using
the correlation of the adjmeanstypes with GEBVs:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>acc_GBLUP <span class="ot">=</span> <span class="fu">cor</span>(adjmeans_L<span class="sc">$</span>Y, GBLUP<span class="sc">$</span>u, <span class="at">use =</span> <span class="st">&#39;complete.obs&#39;</span>);acc_GBLUP</span></code></pre></div>
<pre><code>[1] 0.2896361</code></pre>
<p><br></p>
</div>
<div id="bayesian-ridge-regression" class="section level2">
<h2>Bayesian Ridge Regression</h2>
<p>We can perform the Ridge Regression in the Bayesian framework. For
this purpose, we will use the <a
href="https://github.com/gdlc/BGLR-R">BGLR</a> package <span
class="citation">[@BGLR]</span>. In this package, we will need an
element called “ETA”. In the ETA, we will set the linear predictors of
the model and the priors.</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>ETA <span class="ot">=</span> <span class="fu">list</span>(<span class="fu">list</span>(<span class="at">X =</span> SNPs_L, <span class="at">model =</span> <span class="st">&quot;BRR&quot;</span>))</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>BRR <span class="ot">=</span> <span class="fu">BGLR</span>(<span class="at">y =</span> adjmeans_L<span class="sc">$</span>Y, <span class="at">ETA =</span> ETA, <span class="at">nIter =</span> <span class="dv">15000</span>, <span class="at">burnIn =</span> <span class="dv">5000</span>, </span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">thin =</span> <span class="dv">5</span>, <span class="at">verbose =</span> F)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>BRR_GENO <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="st">&quot;Genotype&quot;</span> <span class="ot">=</span> adjmeans_L<span class="sc">$</span>GENO,</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&quot;Yield&quot;</span> <span class="ot">=</span> adjmeans_L<span class="sc">$</span>Y,</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&quot;GEBV&quot;</span> <span class="ot">=</span> BRR<span class="sc">$</span>yHat)</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>BRR<span class="sc">$</span>ETA[[<span class="dv">1</span>]]<span class="sc">$</span>varB <span class="co">#Variance component of the marker effects</span></span></code></pre></div>
<pre><code>[1] 0.0008386908</code></pre>
<p><br></p>
</div>
<div id="bayes-b" class="section level2">
<h2>Bayes B</h2>
<p>To perform the Bayes B method, we will also use the BGLR package. For
this purpose, we will simply change from “BRR” to “BayesB” in the model
within the ETA element.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>ETA <span class="ot">=</span> <span class="fu">list</span>(<span class="fu">list</span>(<span class="at">X =</span> SNPs_L, <span class="at">model =</span> <span class="st">&quot;BayesB&quot;</span>))</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>BayesB <span class="ot">=</span> <span class="fu">BGLR</span>(<span class="at">y =</span> adjmeans_L<span class="sc">$</span>Y, <span class="at">ETA =</span> ETA, <span class="at">nIter =</span> <span class="dv">15000</span>, <span class="at">burnIn =</span> <span class="dv">5000</span>, </span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">thin =</span> <span class="dv">5</span>, <span class="at">verbose =</span> F)</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>BayesB_GENO <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="st">&quot;Genotype&quot;</span> <span class="ot">=</span> adjmeans_L<span class="sc">$</span>GENO,</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&quot;Yield&quot;</span> <span class="ot">=</span> adjmeans_L<span class="sc">$</span>Y,</span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&quot;GEBV&quot;</span> <span class="ot">=</span> BayesB<span class="sc">$</span>yHat)</span></code></pre></div>
<p><br></p>
<p>We can compare the estimated SNP effects of each method, BRR and
BayesB:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>bhat_BRR <span class="ot">=</span> BRR<span class="sc">$</span>ETA[[<span class="dv">1</span>]]<span class="sc">$</span>b</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>bhat_BayesB <span class="ot">=</span> BayesB<span class="sc">$</span>ETA[[<span class="dv">1</span>]]<span class="sc">$</span>b</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>bhat_BRR_plot <span class="ot">=</span> <span class="fu">as.data.frame</span>(bhat_BRR) <span class="sc">%&gt;%</span> </span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">&quot;SNP&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">BLUP =</span> bhat_BRR) <span class="sc">%&gt;%</span> </span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> SNP, <span class="at">y =</span> BLUP))<span class="sc">+</span></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="dv">0</span>, <span class="at">xend =</span> SNP, <span class="at">yend =</span> BLUP),)<span class="sc">+</span></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>,<span class="fu">max</span>(bhat_BRR))<span class="sc">+</span></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>())<span class="sc">+</span></span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">&quot;Estimated marker effect&quot;</span>, <span class="at">title =</span> <span class="st">&quot;BRR&quot;</span>)</span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>bhat_BayesB_plot <span class="ot">=</span> <span class="fu">as.data.frame</span>(bhat_BayesB) <span class="sc">%&gt;%</span> </span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">&quot;SNP&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">BLUP =</span> bhat_BayesB) <span class="sc">%&gt;%</span> </span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> SNP, <span class="at">y =</span> BLUP))<span class="sc">+</span></span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>()<span class="sc">+</span></span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">y =</span> <span class="dv">0</span>, <span class="at">xend =</span> SNP, <span class="at">yend =</span> BLUP),)<span class="sc">+</span></span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>,<span class="fu">max</span>(bhat_BayesB))<span class="sc">+</span></span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>(), <span class="at">axis.ticks.x=</span><span class="fu">element_blank</span>())<span class="sc">+</span></span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">y=</span><span class="st">&quot;Estimated marker effect&quot;</span>, <span class="at">title =</span> <span class="st">&quot;BayesB&quot;</span>)</span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-24"><a href="#cb38-24" aria-hidden="true" tabindex="-1"></a>bhat_BRR_plot <span class="sc">+</span> bhat_BayesB_plot</span></code></pre></div>
<p><img src="figure/genomicselection.Rmd/unnamed-chunk-25-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="using-eigenvalues---rkhs" class="section level2">
<h2>Using eigenvalues - RKHS</h2>
<p>We can also use the eigenvalues of the <span
class="math inline">\(\mathbf G\)</span> covariance matrix to perform
the analyses. But first, we have to build <span
class="math inline">\(\mathbf G\)</span>:</p>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>genmat <span class="ot">=</span> <span class="fu">model.matrix</span>(<span class="sc">~-</span><span class="dv">1</span><span class="sc">+</span>GENO, <span class="at">data =</span> adjmeans_L)</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>G_matrix_L <span class="ot">=</span> <span class="fu">Gmatrix</span>(SNPs_L, <span class="at">method =</span> <span class="st">&quot;VanRaden&quot;</span>, <span class="at">ploidy =</span> <span class="dv">2</span>, <span class="at">missingValue =</span> <span class="cn">NA</span>)</span></code></pre></div>
<pre><code>Initial data: 
    Number of Individuals: 415 
    Number of Markers: 5000 

Missing data check: 
    Total SNPs: 5000 
     0 SNPs dropped due to missing data threshold of 1 
    Total of: 5000  SNPs 
MAF check: 
    No SNPs with MAF below 0 
Monomorphic check: 
    No monomorphic SNPs 
Summary check: 
    Initial:  5000 SNPs 
    Final:  5000  SNPs ( 0  SNPs removed) 
 
Completed! Time = 3.73  seconds </code></pre>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>G <span class="ot">=</span> <span class="fu">tcrossprod</span>(<span class="fu">tcrossprod</span>(genmat,G_matrix_L),genmat)</span></code></pre></div>
<p>Then, we may estimate the eigenalues (and eigenvectors):</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>EVD_G <span class="ot">=</span> <span class="fu">eigen</span>(G)</span></code></pre></div>
<p>This action is useful for speeding the analyses. After obtaining the
eigenvalues (and the eigenvectors), we can insert them in the ETA
component. In that case, we will the Reproducing Kernel Hilbert Spaces
(RKHS) regressions, which uses the eigenvalues to perform the
analyses</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>ETA <span class="ot">=</span> <span class="fu">list</span>(<span class="fu">list</span>(<span class="at">V =</span> EVD_G<span class="sc">$</span>vectors, <span class="at">d =</span> EVD_G<span class="sc">$</span>values, <span class="at">model =</span> <span class="st">&quot;RKHS&quot;</span>))</span></code></pre></div>
<p>After defining the ETA, we can fit the model:</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>RKHS <span class="ot">=</span> <span class="fu">BGLR</span>(<span class="at">y =</span> adjmeans_L<span class="sc">$</span>Y,<span class="at">ETA =</span> ETA, <span class="at">nIter =</span> <span class="dv">10000</span>, <span class="at">burnIn =</span> <span class="dv">5000</span>, </span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">thin =</span> <span class="dv">5</span>, <span class="at">verbose =</span> F)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>RKHS_GENO <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="st">&quot;Genotype&quot;</span> <span class="ot">=</span> adjmeans_L<span class="sc">$</span>GENO,</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&quot;Yield&quot;</span> <span class="ot">=</span> adjmeans_L<span class="sc">$</span>Y,</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>                      <span class="st">&quot;GEBV&quot;</span> <span class="ot">=</span> RKHS<span class="sc">$</span>yHat)</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>RKHS<span class="sc">$</span>ETA[[<span class="dv">1</span>]]<span class="sc">$</span>varU <span class="co">#Variance component of the genetics effects</span></span></code></pre></div>
<pre><code>[1] 1.75623</code></pre>
<p><br></p>
<p><br></p>
</div>
<div id="cross-validation" class="section level2">
<h2>Cross-validation</h2>
<p>To prove that the prediction is accurate, we should perform a
cross-validation (CV) scheme. For this purpose, we divide the data into
a training set and a validation set. First we separate the data into
<em>k</em> folds. Then, we attribute <em>NA</em> for one fold and try to
predict the data from this fold based on the others. When selecting the
number of folds, one must prioritize the balance between the number of
observations in each fold. In addition, this process should be repeated
for further validation. The step-by-step below will guide the CV in the
data we are analysing.</p>
<div id="determine-the-number-of-folds-and-repetitions"
class="section level3">
<h3>1. Determine the number of folds and repetitions</h3>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>nfolds <span class="ot">=</span> <span class="dv">5</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>nrept <span class="ot">=</span> <span class="dv">2</span></span></code></pre></div>
<p>Since we defined 5 folds, our data will be divided into 5 parts with
83 observations each.</p>
</div>
<div id="match-the-order-of-the-data-and-the-rows-of-the-snp-matrix"
class="section level3">
<h3>2. Match the order of the data and the rows of the SNP matrix</h3>
<p>The order is decreasing or increasing (numeric or alphabetical)
regarding the name of the genotypes.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>adjmeans_L <span class="ot">=</span> adjmeans_L[<span class="fu">order</span>(adjmeans_L<span class="sc">$</span>GENO, <span class="at">decreasing =</span> <span class="cn">FALSE</span>),] </span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>SNPs_L <span class="ot">=</span> SNPs_L[<span class="fu">order</span>(<span class="fu">row.names</span>(SNPs_L)),]</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">rownames</span>(SNPs_L) <span class="sc">==</span> adjmeans_L<span class="sc">$</span>GENO)</span></code></pre></div>
<pre><code>[1] TRUE</code></pre>
</div>
<div id="add-a-column-indicating-a-number-for-each-observation"
class="section level3">
<h3>3. Add a column indicating a number for each observation</h3>
<p>This will be useful to assign each observation for a fold, which will
be the next step.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>adjmeans_L<span class="sc">$</span>ID <span class="ot">=</span> <span class="fu">factor</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(adjmeans_L))</span></code></pre></div>
</div>
<div id="folds-assignment" class="section level3">
<h3>4. Folds assignment</h3>
<p>In this step, we will assign each observation to a fold. Bear in mind
that for each repetition, the folds will comprise different
observations. The purpose of the repetition is to make sure of the
randomness of the assignment step. In this step, we will use the <a
href="https://cran.r-project.org/web/packages/cvTools/cvTools.pdf">cvTools</a>
package <span class="citation">[@cvTools]</span></p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">100</span>)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>sort<span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(a <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nrept){</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nfolds){</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>    folds <span class="ot">&lt;-</span> <span class="fu">cvFolds</span>(<span class="fu">nlevels</span>(adjmeans_L<span class="sc">$</span>ID),<span class="at">type =</span> <span class="st">&quot;random&quot;</span>, <span class="at">K =</span> <span class="dv">5</span>, <span class="at">R =</span> <span class="dv">1</span>)</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>    Sample <span class="ot">&lt;-</span> <span class="fu">cbind</span>(folds<span class="sc">$</span>which,folds<span class="sc">$</span>subsets)</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>    cv <span class="ot">&lt;-</span> <span class="fu">split</span>(Sample[,<span class="dv">2</span>], <span class="at">f=</span>Sample[,<span class="dv">1</span>])</span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>  sort[[a]] <span class="ot">&lt;-</span> cv  </span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a><span class="fu">rm</span>(a, folds, j, cv, Sample)</span></code></pre></div>
</div>
<div id="cross-validation-1" class="section level3">
<h3>5. Cross-validation</h3>
<p>The next step is the very CV. Here, we will define the linear
predictor and the lists that will be useful in the loop. The first list,
here called “fold.list”, contains the folds assignation that we built in
the previous step. The second (“results”) and third (“Out”) lists are
empty and will store the outputs of each iteration of the loop.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>fold.list <span class="ot">=</span> sort</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>results <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>Out <span class="ot">=</span> <span class="fu">list</span>()</span></code></pre></div>
<p>Then, we will construct the loop. Each iteration will assign
<em>NA</em> for a different fold, and we will use the other folds to
predict the missing values. Note that the folds vary for each
repetition.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>ETA <span class="ot">=</span> <span class="fu">list</span>(<span class="fu">list</span>(<span class="at">X =</span> SNPs_L, <span class="at">model =</span> <span class="st">&quot;BayesB&quot;</span>))</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (z <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(fold.list)) {</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nfolds){</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Training set</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>    train_data <span class="ot">&lt;-</span> adjmeans_L </span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Validation set</span></span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>    train_data[train_data<span class="sc">$</span>ID <span class="sc">%in%</span> fold.list[[z]][[i]], <span class="st">&quot;Y&quot;</span>] <span class="ot">&lt;-</span> <span class="cn">NA</span> </span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Fitting model </span></span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a>    BB_M <span class="ot">&lt;-</span> <span class="fu">BGLR</span>(<span class="at">y =</span> train_data<span class="sc">$</span>Y, <span class="at">ETA =</span> ETA, <span class="at">nIter =</span> <span class="dv">10000</span>, <span class="at">burnIn =</span> <span class="dv">5000</span>, </span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a>                 <span class="at">thin =</span> <span class="dv">5</span>, <span class="at">verbose =</span> F)</span>
<span id="cb52-14"><a href="#cb52-14" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb52-15"><a href="#cb52-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># GEBV</span></span>
<span id="cb52-16"><a href="#cb52-16" aria-hidden="true" tabindex="-1"></a>    Pred <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Yhat =</span> BB_M<span class="sc">$</span>yHat, <span class="at">G =</span> adjmeans_L<span class="sc">$</span>ID)</span>
<span id="cb52-17"><a href="#cb52-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(Pred) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(SNPs_L)</span>
<span id="cb52-18"><a href="#cb52-18" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb52-19"><a href="#cb52-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Predicted GEBV</span></span>
<span id="cb52-20"><a href="#cb52-20" aria-hidden="true" tabindex="-1"></a>    results[[i]] <span class="ot">&lt;-</span> Pred[Pred[, <span class="st">&quot;G&quot;</span>] <span class="sc">%in%</span> fold.list[[z]][[i]], ] </span>
<span id="cb52-21"><a href="#cb52-21" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb52-22"><a href="#cb52-22" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Remove</span></span>
<span id="cb52-23"><a href="#cb52-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rm</span>(BB_M, Pred, train_data)</span>
<span id="cb52-24"><a href="#cb52-24" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb52-25"><a href="#cb52-25" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-26"><a href="#cb52-26" aria-hidden="true" tabindex="-1"></a>  GEBV <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, results)</span>
<span id="cb52-27"><a href="#cb52-27" aria-hidden="true" tabindex="-1"></a>  GEBV <span class="ot">&lt;-</span> GEBV[<span class="fu">order</span>(GEBV<span class="sc">$</span>G), ]</span>
<span id="cb52-28"><a href="#cb52-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-29"><a href="#cb52-29" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Log</span></span>
<span id="cb52-30"><a href="#cb52-30" aria-hidden="true" tabindex="-1"></a>  log <span class="ot">&lt;-</span> <span class="fu">all</span>(GEBV<span class="sc">$</span>G <span class="sc">==</span> adjmeans_L<span class="sc">$</span>ID)</span>
<span id="cb52-31"><a href="#cb52-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-32"><a href="#cb52-32" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Results</span></span>
<span id="cb52-33"><a href="#cb52-33" aria-hidden="true" tabindex="-1"></a>  Out[[z]] <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb52-34"><a href="#cb52-34" aria-hidden="true" tabindex="-1"></a>    <span class="at">Rep =</span> z,</span>
<span id="cb52-35"><a href="#cb52-35" aria-hidden="true" tabindex="-1"></a>    <span class="at">Log =</span> log,</span>
<span id="cb52-36"><a href="#cb52-36" aria-hidden="true" tabindex="-1"></a>    <span class="at">Ac =</span> <span class="fu">round</span>(<span class="fu">cor</span>(GEBV<span class="sc">$</span>Yhat, adjmeans_L<span class="sc">$</span>Y, <span class="at">use =</span> <span class="st">&quot;na.or.complete&quot;</span>), <span class="dv">3</span>),</span>
<span id="cb52-37"><a href="#cb52-37" aria-hidden="true" tabindex="-1"></a>    <span class="at">MSPE =</span> <span class="fu">round</span>(<span class="fu">mean</span>(((GEBV<span class="sc">$</span>Yhat <span class="sc">-</span> adjmeans_L<span class="sc">$</span>Y)<span class="sc">^</span><span class="dv">2</span>), <span class="at">na.rm =</span> T), <span class="dv">3</span>)</span>
<span id="cb52-38"><a href="#cb52-38" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb52-39"><a href="#cb52-39" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb52-40"><a href="#cb52-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-41"><a href="#cb52-41" aria-hidden="true" tabindex="-1"></a>Out</span></code></pre></div>
<pre><code>[[1]]
  Rep  Log    Ac MSPE
1   1 TRUE 0.213 8.51

[[2]]
  Rep  Log    Ac  MSPE
1   2 TRUE 0.208 8.541</code></pre>
<p>The object “Out” is divided by repetition. In the “Out” objects for
each repetition, “Rep” is the number of the repetition, “Log” is a
diagnostic indicating if the order of the predicted breeding values
matches the order of the adjusted means, “Ac” is the prediction accuracy
(correlation between the GEBV and adjusted means), and “MSPE” is the
mean square prediction error (the lower, the better).</p>
<p>The CV is the last step in the single-environment genomic
selection/prediction. These models can be expanded for a
multi-environment context.</p>
</div>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>R version 4.1.3 (2022-03-10)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 19042)

Matrix products: default

locale:
[1] LC_COLLATE=Portuguese_Brazil.1252  LC_CTYPE=Portuguese_Brazil.1252   
[3] LC_MONETARY=Portuguese_Brazil.1252 LC_NUMERIC=C                      
[5] LC_TIME=Portuguese_Brazil.1252    

attached base packages:
[1] grid      stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] patchwork_1.1.2       cvTools_0.3.2         robustbase_0.95-0    
 [4] lattice_0.20-45       ComplexHeatmap_2.10.0 forcats_0.5.2        
 [7] stringr_1.4.1         dplyr_1.0.10          purrr_0.3.4          
[10] readr_2.1.2           tidyr_1.2.1           tibble_3.1.8         
[13] ggplot2_3.3.6         tidyverse_1.3.2       BGLR_1.1.0           
[16] rrBLUP_4.6.1          AGHmatrix_2.0.4       data.table_1.14.2    
[19] RCurl_1.98-1.8        kableExtra_1.3.4     

loaded via a namespace (and not attached):
 [1] googledrive_2.0.0   colorspace_2.0-3    rjson_0.2.21       
 [4] ellipsis_0.3.2      rprojroot_2.0.3     circlize_0.4.15    
 [7] GlobalOptions_0.1.2 fs_1.5.2            clue_0.3-61        
[10] rstudioapi_0.14     farver_2.1.1        fansi_1.0.3        
[13] lubridate_1.8.0     xml2_1.3.3          codetools_0.2-18   
[16] doParallel_1.0.17   cachem_1.0.6        knitr_1.40         
[19] jsonlite_1.8.0      workflowr_1.7.0     broom_1.0.1        
[22] cluster_2.1.2       dbplyr_2.2.1        png_0.1-7          
[25] compiler_4.1.3      httr_1.4.4          backports_1.4.1    
[28] assertthat_0.2.1    Matrix_1.5-1        fastmap_1.1.0      
[31] gargle_1.2.1        cli_3.3.0           later_1.3.0        
[34] htmltools_0.5.3     tools_4.1.3         gtable_0.3.1       
[37] glue_1.6.2          Rcpp_1.0.9          cellranger_1.1.0   
[40] jquerylib_0.1.4     vctrs_0.4.1         svglite_2.1.0      
[43] iterators_1.0.14    xfun_0.32           rvest_1.0.3        
[46] lifecycle_1.0.3     googlesheets4_1.0.1 DEoptimR_1.0-11    
[49] MASS_7.3-58.1       zoo_1.8-11          scales_1.2.1       
[52] hms_1.1.2           promises_1.2.0.1    parallel_4.1.3     
[55] RColorBrewer_1.1-3  yaml_2.3.5          curl_4.3.3         
[58] sass_0.4.2          stringi_1.7.6       highr_0.9          
[61] S4Vectors_0.32.4    foreach_1.5.2       BiocGenerics_0.40.0
[64] truncnorm_1.0-8     shape_1.4.6         rlang_1.0.6        
[67] pkgconfig_2.0.3     systemfonts_1.0.4   bitops_1.0-7       
[70] matrixStats_0.62.0  evaluate_0.17       labeling_0.4.2     
[73] tidyselect_1.2.0    magrittr_2.0.3      R6_2.5.1           
[76] magick_2.7.3        IRanges_2.28.0      generics_0.1.3     
[79] DBI_1.1.3           pillar_1.8.1        haven_2.5.1        
[82] withr_2.5.0         modelr_0.1.9        crayon_1.5.2       
[85] utf8_1.2.2          tzdb_0.3.0          rmarkdown_2.17     
[88] GetoptLong_1.0.5    readxl_1.4.1        git2r_0.30.1       
[91] reprex_2.0.2        digest_0.6.29       webshot_0.5.4      
[94] httpuv_1.6.5        stats4_4.1.3        munsell_0.5.0      
[97] viridisLite_0.4.1   bslib_0.4.0        </code></pre>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr />
<ol>
<li id="fn1"><p>Weverton Gomes da Costa, Pós-Doutorando, Embrapa
Mandioca e Fruticultura, <a href="mailto:wevertonufv@gmail.com"
class="email">wevertonufv@gmail.com</a><a href="#fnref1"
class="footnote-back">↩︎</a></p></li>
</ol>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
